### ISSUE TITLE
Provide GitHub Action to rollback a release

### ISSUE THREAD

**Issue by cpswan** [atsign-foundation/at_server#706]
**Is your feature request related to a problem? Please describe.**
When a release is found to have bugs we need a mechanism to safely roll back to an earlier known good release.

**Describe the solution you'd like**
Workflow that tags an older release as :prod and :canary

**Comment by kumarnarendra701** [comment]
@cpswan - Worked on POC of GitHub action rollback process and found two approaches for that. Please review and share your thoughts.

1. Revert to the previous tag - Delete the failed tag and workflow will pick the last tag and deploy it.
```
name: Deploy and rollback
on:
  push:
  delete:
    tags:
      - 'v*.*.*'
jobs:
  deploy_release:
    runs-on: ubuntu-latest

    steps:
      - id: latest
        uses: thebritican/fetch-latest-release@v1.0.3
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}

      - uses: actions/checkout@v2
        with: 
          ref: ${{ steps.latest.outputs.tag_name }}
      
      - name: Deploy latest release Code
        run: cat test.txt
```
2. Revert to the previous release  - Delete the failure release and workflow will pick the last successful release and deploy it.

```
on: 
  release:
    types: [published, deleted]
    
name: Deploy or Rollback
jobs:
  deploy_release:
    runs-on: ubuntu-latest

    steps:
      - id: latest
        uses: thebritican/fetch-latest-release@v1.0.3
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}

      - uses: actions/checkout@v2
        with: 
          ref: ${{ steps.latest.outputs.tag_name }}
      
      - name: Deploy latest release Code
        run: cat test.txt
```

**Comment by cpswan** [comment]
@kumarnarendra701 this should be about deploys rather than releases. So we're making our infrastructure (Shepherd) deploy a previous release.

We should not be deleting releases.

The appropriate trigger would be manual (e.g. Workflow dispatch).

What the action needs to do is rebuild Docker image manifests so that the key tags (e.g. `:prod` point to images that were previously released). The input to the workflow dispatch should capture which release to go back to (and if that can be a drop down that would be great).

**Comment by kumarnarendra701** [comment]
@cpswan - In Workflow dispatch, we have the option to select tags and trigger actions for tags.
![image](https://user-images.githubusercontent.com/29299536/175037461-0f1a2821-9193-451d-8699-d8fd4638792d.png)

**Comment by kumarnarendra701** [comment]
@cpswan - Please let me know from which repository need to start.

**Comment by cpswan** [comment]
@kumarnarendra701 this repo - at_server
and for atsigncompany/secondary image on Docker Hub

**Comment by kumarnarendra701** [comment]
@cpswan - Added rollback workflow and raised PR for review. 

In PR, I have added a select button to roll back particular steps, and based on selection it will trigger Job steps required steps. If you want to ignore all test cases during rollback in that case we can add one more steps to run only docker build and push. Please suggest accordingly.

```
  # This job run's by workflow_dispatch i.e Manual run.
  # The job builds the production version of secondary server docker image and pushes to docker hub on rollback.
  rollback_prod_secondary_image:
    **needs: functional_tests --------> We will run only functional_tests because we need at_secondary_server binary**
    if: github.event.inputs.rollback_prod_secondary_image == 'true'
    env:
      working-directory: at_server
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3.0.2

      # Extract version for docker tag
      - name: Get version
        run: echo "VERSION=${GITHUB_REF##*/}" >> $GITHUB_ENV

      - name: Set up QEMU
        uses: docker/setup-qemu-action@v2.0.0

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v2.0.0

      - name: Login to DockerHub
        uses: docker/login-action@v2.0.0
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}

      # Gets the secondary server binary generated from the above "secondary" job.
      - name: Get secondary server
        uses: actions/download-artifact@v3
        with:
          name: secondary-server
          path: at_secondary/at_secondary_server

      # Builds and pushes the secondary server image to docker hub.
      - name: Build and push secondary image for amd64 and arm64
        id: docker_build_secondary
        uses: docker/build-push-action@v3.0.0
        with:
          push: true
          context: at_secondary
          tags: |
            atsigncompany/secondary:prod
            atsigncompany/secondary:prod-${{ env.VERSION }}
          platforms: |
            linux/amd64
            linux/arm64/v8

      - name: Create dess-x64 label
        run: |
          docker pull atsigncompany/secondary:prod
          docker tag atsigncompany/secondary:prod atsigncompany/secondary:dess-x64
          docker push atsigncompany/secondary:dess-x64

      - name: Create dess-arm64 label
        uses: appleboy/ssh-action@v0.1.4
        with:
          host: "arm64cicd.atsign.wtf"
          username: ubuntu
          key: ${{ secrets.CICD_SSH_KEY }}
          script: |
            docker pull atsigncompany/secondary:prod
            docker tag atsigncompany/secondary:prod atsigncompany/secondary:dess-arm64
            docker push atsigncompany/secondary:dess-arm64

      # Logs into Pi and builds Armv7 dess secondary
      - name: Create dess-arm image
        uses: appleboy/ssh-action@v0.1.4
        with:
          host: "arm32cicd.atsign.wtf"
          username: pi
          key: ${{ secrets.CICD_SSH_KEY }}
          script: |
            cd /home/pi/git/github.com/atsign-foundation/at_server
            git pull
            cd at_secondary
            docker build -t atsigncompany/secondary:dess-arm .
            docker push atsigncompany/secondary:dess-arm

      - name: Create and push multi-arch dess manifest
        run: |
          docker manifest create atsigncompany/secondary:dess \
            --amend atsigncompany/secondary:dess-arm \
            --amend atsigncompany/secondary:dess-arm64 \
            --amend atsigncompany/secondary:dess-x64
          docker manifest push atsigncompany/secondary:dess
```

**Comment by cpswan** [comment]
Bumped priority to Med, as this is needed by #750, which is prioritised as Med

**Comment by kumarnarendra701** [comment]
@cpswan - As suggested modified workflow, please review and share your feedback.

**Comment by cpswan** [comment]
Still not quite done with this, so moving to PR41

**Comment by kumarnarendra701** [comment]
@cpswan - As suggested added final steps in workflow for rollback. Please review the PR.
https://github.com/atsign-foundation/at_server/pull/779

**Comment by cpswan** [comment]
Review comments added. I'm not sure if #779 should be merged given that similar changes are also in #817.

**Comment by kumarnarendra701** [comment]
@cpswan - Please let me know if any work is required in this ticket or can we close this?
